src/crypto/der.ts

```diff
@@ -0,0 +1,117 @@
+export type AsnSeq = { seq: AsnType[] }
+export type AsnInt = { int: number }
+export type AsnBytes = { bytes: Uint8Array }
+export type AsnIdent = { ident: string }
+export type AsnNull = null;
+
+export type AsnType = AsnSeq | AsnInt | AsnBytes | AsnIdent | AsnNull;
+
+/**
+ * Note: may throw weird errors on malformed input. Catch and rethrow with, e.g. `BadKeyError`.
+ */
+export function decode(derBytes: Uint8Array): AsnType {
+    const [ asn ] = decodeIncremental(derBytes);
+    return asn;
+}
+
+export function decodeIncremental(bytes: Uint8Array): [AsnType, Uint8Array] {
+    // slice off the initial tag byte, `decodeLength` returns a slice of the remaining data
+    const [ len, rem ] = decodeLength(bytes.subarray(1));
+    const data = rem.subarray(0, len);
+    const tail = rem.subarray(len);
+
+    switch (bytes[ 0 ]) {
+        case 2:
+            return [{ int: decodeInt(data) }, tail ];
```

Comment:
The asymmetry in the spacing rules here really bugs me and there's no way to configure IntelliJ's autoformat to fix this.